### ISSUE TITLE
Asking Agda to solve a constraint inside a macro

### ISSUE THREAD

**Issue by nad** [agda/agda#3791]
Consider the following code:
```agda
module _ where

open import Agda.Builtin.Bool
open import Agda.Builtin.List
open import Agda.Builtin.Reflection
open import Agda.Builtin.Unit

record R : Set₁ where
  field
    _≡_  : {A : Set} → A → A → Set
    refl : {A : Set} (x : A) → x ≡ x

record R′ (_ : Set) : Set where

module _ (r : R) (_ : {A : Set} → R′ A) where

  open R r

  macro

    m : Term → TC ⊤
    m goal =
      bindTC (unify (def (quote refl)
                         (arg (arg-info visible relevant) unknown ∷ []))
                    goal) λ _ →
      bindTC (reduce goal) λ where
        (meta _ _) → typeError (strErr "Meta" ∷ [])
        _          → returnTC _

  test : true ≡ true
  test = m
```
This code is rejected:
```
Meta
when checking that the expression unquote m has type true ≡ true
```
If `typeError (strErr "Meta" ∷ [])` is replaced by `returnTC _`, then the code is accepted. This suggests to me that there is some constraint that can be solved, but isn't solved by the call to `unify`. Is there some way to get Agda to solve this constraint inside the macro? If not, should we make it possible to do this?

**Comment by jespercockx** [comment]
I'm in favour of adding a `solveConstraints` primitive to the reflection API. In the past I encountered some similar problems which I managed to solve by creative use of `catchError` and `blockOnMeta`, but being able to just call the constraint solver directly would've been much easier.

Edit: maybe the primitive should take an argument of type `MetaId -> Bool` to only solve constraints that mention certain metavariables.

**Comment by nad** [comment]
> maybe the primitive should take an argument of type `MetaId -> Bool` to only solve constraints that mention certain metavariables.

Yes.

**Comment by nad** [comment]
Is there currently some efficient way to find out which constraints mention a certain set of meta-variables?

**Comment by jespercockx** [comment]
I don't know about efficiency, but the current code implementing this internally can be found here:

https://github.com/agda/agda/blob/6c7f3ff048ac18c3711eff9a26353fde51dbefb7/src/full/Agda/TypeChecking/Constraints.hs#L175-L192

(edit: fixed link)

**Comment by nad** [comment]
OK, not so efficient.

Does anyone know if "`isConsideringInstance`" can be true when the new primitive(s) are called from a macro? I want to know whether this detail should be explained in the documentation.

**Comment by jespercockx** [comment]
As the one who added it, I'm pretty sure that `isConsideringInstance` is only true during the candidate check of instance search, so it is not something users of the reflection API should have to worry about.

**Comment by nad** [comment]
@jespercockx and I noticed that the new primitive `solveConstraints` does not solve constraints if Agda is already solving constraints when the tactic is run. We tried to change the implementation so that constraints are solved anyway, but this lead to a recursive invocation of a tactic. This problem should be addressed in some way before the new primitives are released. Some possible ways forward:
* Remove the new primitives.
* Document this behaviour.
* Switch to some kind of predictable implementation that, for instance, tries to solve all constraints except the ones that Agda is currently trying to solve (perhaps also removing some or all of these constraints from pending work lists).
